Add status-provider Flag

suggoitanoshi opened the pull request:
This adds a status-provider flag to mix and match between Status
Providers and Hosts, while falling back to defaults.

Closes https://github.com/ksdme/org-status/issues/17

kx-chen commented on org_status/org_status.py:
```diff
@@ -87,6 +87,7 @@ def get_argument_parser():
     parser.add_argument('--export-repos', type=str)
     parser.add_argument('--format', type=str, default='gitman')
     parser.add_argument('--check-providers-only', action='store_true')
+    parser.add_argument('--status-provider', type=str)
```
`--status-providers`? I think it'd be more clear.

kx-chen commented on org_status/org_status.py:
```diff
@@ -36,7 +36,7 @@ def generate_fetch_jobs(org_strings):
                 if host == supported_host.HostName:
                     yield (supported_host, org)
-                    raise StopIteration
+                    # raise StopIteration
```
Either get rid of the code, or leave it in, but please don't have it as a comment.

kx-chen commented on org_status/org_hosts/gitlab.py:
```diff
@@ -35,10 +38,29 @@ def process_repository(self, repo, branch='master'):
-                                                       branch=branch)
-
+        repo_status = []
+        for i in enumerate(self._status_provider):
```
Originally, the hacky `if` statement was added in https://github.com/ksdme/org-status/pull/19 as a temporary solution. This isn't the ideal way to check each StatusProvider.

Please see https://github.com/ksdme/org-status/pull/19#discussion_r230565364 (and the other comments on there) for ideas on how to do it.

kx-chen commented on org_status/org_hosts/gitlab.py:
```diff
@@ -35,10 +38,29 @@ def process_repository(self, repo, branch='master'):
+            if Status.PASSING in repo_status:
+                repo_status = Status.PASSING
+            # if statuses are identical, just return one of them
+            elif repo_status[0] == repo_status[1]:
```
This `if` statement will fail if there is more than one result. The entire `if` is not a good solution; I think it needs to be re thought.

kx-chen commented on org_status/org_status.py:
```diff
@@ -146,6 +147,22 @@ def main():
+                    status_provider.append(supported_provider)
+                    found = True
+                    break
+            if found:
```
Why not just check for `if not found`?

kx-chen commented on org_status/org_hosts/github.py:
```diff
@@ -42,19 +42,23 @@ def process_repository(self, repo, branch='master'):
-            repo_status = Status.PASSING
-        # if statuses are identical, just return one of them
-        elif repo_status[0] == repo_status[1]:
+        if len(repo_status) > 1:
```
We shouldn't be repeating the same code again in `gitlab.py`. See other comments for suggestions instead of these `if`'s.

kx-chen reviewed (commented):
Left some comments :)

ksdme reviewed (commented):
I think the best way to implement this functionality would be to simply extend syntax to submit a task to include the status provider to use, say to something like `travis@gitlab:coala` or some other proper alternate instead of just `gitlab:coala`. We can prevent these kinds of hacky solutions by simply ensuring that an instance of OrgHost is only attached to one StatusProvider which can be configured during object construction. This way we can integrate well with the current infrastructure and not break anything.

suggoitanoshi committed:
Add status-provider in org argument

This commit adds a pattern to determine status-provider. The
status-provider is determined by @ at the beginning of `orgs` argument

Closes https://github.com/ksdme/org-status/issues/17

ksdme commented:
Requires more work, please be on stand by. I'll review it a bit later.

suggoitanoshi commented:
Alright, for now I'll assess my code once more and see what I can do.

On Tue, Nov 27, 2018, 20:51 Kilari Teja <notifications@github.com wrote:

> Requires more work, please be on stand by. I'll review it a bit later.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ksdme/org-status/pull/38#issuecomment-442064856>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKF4UrnIZTjYHYSwhej3NhrQMBQxf9Vmks5uzUNOgaJpZM4YxwW7>
> .
>
